modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+  double edge_point1[3] = {p1(0), p1(1), p1(2)};
+  double edge_point2[3] = {p2(0), p2(1), p2(2)};
+
+  double cut_point[3] = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(
+          plane_point, planenormal, edge_point1, edge_point2, cut_point) == 1)
+  {
+    Point temp_p(cut_point[0], cut_point[1], cut_point[2]);
+    if (isInsideCutPlane(vertices, temp_p) && isInsideEdge(p1, p2, temp_p))
+    {
+      pint = temp_p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::findIntersection(const Point & p1,
+                                      const Point & p2,
+                                      const std::vector<Point> & vertices,
+                                      Point & pint) const
+{
+  // find directional intersection
+  // along the positive extension of the vector from p1 to p2
+  bool has_intersection = false;
+
+  Point normal;
+  Point center;
+  findElemNormal(vertices, normal, center);
+
+  double plane_point[3] = {center(0), center(1), center(2)};
+  double planenormal[3] = {normal(0), normal(1), normal(2)};
+  double p_begin[3] = {p1(0), p1(1), p1(2)};
+  double p_end[3] = {p2(0), p2(1), p2(2)};
+
+  double cut_point[3] = {0.0, 0.0, 0.0};
+
+  if (Xfem::plane_normal_line_exp_int_3d(plane_point, planenormal, p_begin, p_end, cut_point) == 1)
+  {
+    Point p(cut_point[0], cut_point[1], cut_point[2]);
+    Real dotp = ((p - p1) * (p2 - p1)) / ((p2 - p1) * (p2 - p1));
+    if (isInsideCutPlane(vertices, p) && dotp > 1)
+    {
+      pint = p;
+      has_intersection = true;
+    }
+  }
+  return has_intersection;
+}
+
+bool
+MeshCut3DUserObject::isInsideEdge(const Point & p1, const Point & p2, const Point & p) const
+{
+  Real dotp1 = (p1 - p) * (p2 - p1);
+  Real dotp2 = (p2 - p) * (p2 - p1);
+  return (dotp1 * dotp2 <= 0.0);
+}
+
+Real
+MeshCut3DUserObject::getRelativePosition(const Point & p1, const Point & p2, const Point & p) const
+{
+  // get the relative position of p from p1
+  Real full_len = (p2 - p1).norm();
+  Real len_p1_p = (p - p1).norm();
+  return len_p1_p / full_len;
+}
+
+bool
+MeshCut3DUserObject::isInsideCutPlane(const std::vector<Point> & vertices, Point p) const
+{
+  unsigned int n_node = vertices.size();
+
+  Point normal;
+  Point center;
+
+  findElemNormal(vertices, normal, center);
+
+  bool inside = false;
+  unsigned int counter = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1 = (i < n_node - 1 ? i + 1 : 0);
+    Point middle2p = p - 0.5 * (vertices[i] + vertices[iplus1]);
+    const Point side_tang = vertices[iplus1] - vertices[i];
+    Point side_norm = side_tang.cross(normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+bool
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
```

Comment:
Just a normal from a face? Unless I'm missing something, you should be able to do this already with a normal libMesh `Elem`.